libvast/src/system/get_command.cpp

```diff
@@ -0,0 +1,120 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include "vast/system/get_command.hpp"
+
+#include "vast/aliases.hpp"
+#include "vast/concept/parseable/to.hpp"
+#include "vast/concept/parseable/vast/si.hpp"
+#include "vast/defaults.hpp"
+#include "vast/detail/assert.hpp"
+#include "vast/detail/make_io_stream.hpp"
+#include "vast/format/ascii.hpp"
+#include "vast/format/json.hpp"
+#include "vast/fwd.hpp"
+#include "vast/ids.hpp"
+#include "vast/logger.hpp"
+#include "vast/scope_linked.hpp"
+#include "vast/system/archive.hpp"
+#include "vast/system/node_control.hpp"
+#include "vast/system/spawn_or_connect_to_node.hpp"
+
+#include <caf/expected.hpp>
+#include <caf/scoped_actor.hpp>
+#include <caf/settings.hpp>
+
+#include <vector>
+
+namespace vast::system {
+
+namespace {
+
+ids to_ids(const count& x) {
+  ids result;
+  result.append_bits(false, x);
+  result.append_bit(true);
+  return result;
+}
+
+template <class Writer>
+caf::error
+run(caf::scoped_actor& self, archive_type& archive, const invocation& inv) {
+  auto make_writer = [&]() -> caf::expected<Writer> {
+    using ostream_ptr = std::unique_ptr<std::ostream>;
+    if constexpr (std::is_constructible_v<Writer, ostream_ptr>) {
+      auto output = get_or(inv.options, "get.write", "-");
+      auto uds = get_or(inv.options, "get.uds", false);
+      auto out = detail::make_output_stream(output, uds);
+      if (!out)
+        return {out.error()};
+      return Writer{std::move(*out)};
+    } else {
+      return Writer{};
+    }
+  };
+  auto writer = make_writer();
+  if (!writer)
+    return writer.error();
+  // TODO: Sending one id at a time is overly pessimistic. A smarter algorithm
+  // would request all ids at once and reorder the results for printing.
```

Comment:
While a valid point, the idea behind this command was to figure out possible performance issues with the archive, and I think doing this 1 by 1 is fine for this purpose. Maybe it should just be documented that this command is not optimized for performance.

Reply:
It depends, I would argue that many queries generate contiguous id ranges, but the mechanism here would break them apart. This is useful to analyze worst-case situations, but it can't be used to simulate normal workloads.